### ISSUE TITLE
`invalid memory address or nil pointer dereference` when using `.Page.File.Dir` in a partial

### ISSUE THREAD

**Issue by stockholmux** [gohugoio/hugo#12221]
### What version of Hugo are you using (`hugo version`)?

```
hugo v0.123.8-5fed9c591b694f314e5939548e11cc3dcb79a79c+extended darwin/arm64 BuildDate=2024-03-07T13:14:42Z VendorInfo=brew
```

### Does this issue reproduce with the latest release?

(I am running the latest release).

I have a partial that that is called from another partial (if relevant, [docsy](https://github.com/google/docsy) template `head.html` calling `head-end.html`).

In the partial, hugo chokes on the following line:

```
{{- $file := print .Page.File.Dir  -}}
```

Yielding the following error message:

```
...execute of template failed: template: partials/hooks/head-end.html:2:24: executing "partials/hooks/head-end.html" at <.Page.File.Dir>: error calling Dir: runtime error: invalid memory address or nil pointer dereference
```

The same tempate worked fine on hugo `v0.104.2`. 

On v0.123.8, I was able to work around this by parsing `.Page` in a rather nasty way:

```
{{- $file := strings.Replace (strings.Replace (print .Page) "Page(/" "") ")" "" -}}
```

Interestingly, I'm using `.Page.File.Dir` in a shortcode on v0.123.8 and it's not throwing an error.

**Comment by jmooring** [comment]
Some pages are not backed by a file, so calling a `.File` method is nonsensical. Code defensively:

```text
{{ with .File })
  {{ .SomeFileMethod }}
{{ end }}
```

It works in a shortcode because shortcodes are part of content which, by definition, is backed by a page.

**Comment by stockholmux** [comment]
@jmooring Okay, why did it work in previous versions?

Additionally, the directory is available from `print .Page`? 

This certainly seems like a bug.

**Comment by jmooring** [comment]
Place this in your baseof.html file:

```text
{{ .File.Dir }}
```

Build with v0.122.0

> WARN  .File.Dir on zero object. Wrap it in if or with: {{ with .File }}{{ .Dir }}{{ end }}

We've been warning users about this for years, and have finally removed the warning in favor of an error:

> Error: error building site: render: failed to render pages: render of "term" failed: "/home/jmooring/code/hugo-testing/layouts/_default/baseof.html:17:10": execute of template failed: template: _default/list.html:17:10: executing "_default/list.html" – File is nil; wrap it in if or with: {{ with .File }}{{ .Dir }}{{ end }}


And `print .Page` is also nonsensical... you're printing the string representation of a page object. Perhaps you want `.Page.Path`, which gives you the logical path to the page, which may or may not be backed by a file.

**Comment by stockholmux** [comment]
I indeed may be able to use `.Page.Path` instead of `.Page.File.Dir`.

A few things on this issue @jmooring:

1. The [current documentation doesn't reflect the error vs warning change - it still says a warning](https://gohugo.io/methods/page/file/#defensive-coding). It currently says "Without a backing file, Hugo will throw a **warning** if you attempt to access a .File property." (emphasis on warning is mine).

2. `invalid memory address or nil pointer dereference` is probably an inappropriate error message here. Before 0.123.0, you got a simple warning and no other error messages. After 0.123.0, I got pages of `invalid memory address or nil pointer dereference` errors  _before_ a single `File is nil; wrap it in if or with`. Debugging wise, I tried to address the first error first.  If these pages of `invalid memory address or nil pointer dereference` were suppressed and the real error was displayed, I probably wouldn't have been stumped nor had to write this issue.
